osquery/tables/system/linux/md_tables.cpp

```diff
@@ -0,0 +1,733 @@
+  std::unique_ptr<udev_enumerate, decltype(delUdevEnum)> udevEnum(
+      udev_enumerate_new(handle.get()), delUdevEnum);
+  if (udevEnum.get() == nullptr) {
+    LOG(ERROR) << "Could not get enumerate handle\n";
+    return;
+  }
+
+  udev_enumerate_add_match_subsystem(udevEnum.get(), systemName.c_str());
+  udev_enumerate_scan_devices(udevEnum.get());
+  udev_list_entry* device_entries =
+      udev_enumerate_get_list_entry(udevEnum.get());
+
+  udev_list_entry* entry;
+  udev_list_entry_foreach(entry, device_entries) {
+    const char* path = udev_list_entry_get_name(entry);
+
+    auto delUdevDevice = [](udev_device* d) { udev_device_unref(d); };
+    std::unique_ptr<udev_device, decltype(delUdevDevice)> device(
+        udev_device_new_from_syspath(handle.get(), path), delUdevDevice);
+    if (device.get() == nullptr) {
+      LOG(ERROR) << "Could not get udev device handle\n";
+      continue;
+    }
+    if (f(device.get())) {
+      break;
+    }
+  }
+}
+
+std::string MD::getPathByDevName(std::string name) {
+  std::string devPath;
+
+  walkUdevDevices("block", [&](udev_device* const& device) {
+    const char* devName = udev_device_get_property_value(device, "DEVNAME");
+
+    if (strcmp(name.c_str(), &devName[strlen(devName) - name.length()]) == 0) {
+      devPath = devName;
+      if (devPath.find('/') != 0) {
+        devPath = "/dev/" + devPath;
+      }
+
+      return true;
+    }
+
+    return false;
+  });
+
+  return devPath;
+}
+
+std::string MD::getDevName(int major, int minor) {
+  std::string devName = "unknown";
+
+  walkUdevDevices("block", [&](udev_device* const& device) {
+    const char* devMajor = udev_device_get_property_value(device, "MAJOR");
+    const char* devMinor = udev_device_get_property_value(device, "MINOR");
+
+    if (std::stoi(devMajor) == major && std::stoi(devMinor) == minor) {
+      devName = udev_device_get_property_value(device, "DEVNAME");
+      return true;
+    }
+
+    return false;
+  });
+
+  return devName;
+}
+
+std::string MD::getSuperblkVersion(std::string arrayName) {
+  std::string version;
+
+  walkUdevDevices("block", [&](udev_device* const& device) {
+    const char* devName = udev_device_get_property_value(device, "DEVNAME");
+
+    if (strcmp(arrayName.c_str(),
+               &devName[strlen(devName) - arrayName.length()]) == 0) {
+      version = udev_device_get_property_value(device, "MD_METADATA");
+      return true;
+    }
+
+    return false;
+  });
+
+  return version;
+}
+
+/**
+ * @brief resolves MD disk state field to string representation
+ *
+ * @param  state state field of mdu_disk_info_t
+ *
+ * @return stringified state
+ */
+std::string getDiskStateStr(int state) {
+  // If state is 0, which is undefined, we assume recoverying, as this is all
+  // have seen in the wild
+  if (state == 0)
+    return "recovering";
+
+  std::string s;
+
+  if ((1 << MD_DISK_FAULTY) & state)
+    s += "faulty ";
+
+  if ((1 << MD_DISK_ACTIVE) & state)
```

Comment:
Might this be better suited by a `switch (state)` `case ...` as opposed to a series of `if`s?